File: sdk-python/gatedhouse/_login_flow.py
"""Login-CSRF-safe hosted-login flow for Sphinx (mirrors Java ``LoginFlow``).

Binds the authorization code to the browser that started the flow using **PKCE**
(not ``state``): ``begin_login`` mints a PKCE ``code_verifier``, stashes it in a
signed cookie value bound to that browser, and returns the ``/oauth/authorize``
URL; ``complete_login`` requires that cookie and redeems the code *with* the
verifier, so a code minted for a different browser's flow is rejected by Sphinx
before the app adopts any identity.

Framework-agnostic: unlike the Java SDK (which reads/writes servlet cookies
directly), this returns cookie *values* for the host to set/read with its own
web framework. Set the ``gh_login`` cookie ``HttpOnly``, ``Secure``,
``SameSite=Lax``, ``Max-Age = COOKIE_MAX_AGE_SECONDS``; on
``complete_login`` clear it and rotate the session id on elevation.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import os
import urllib.parse

from ._exceptions import LoginCsrfError
from ._secure_urls import require_https_or_loopback
from ._sphinx_client import SphinxClient, TokenResponse


def _b64url(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


class LoginFlow:

    #: Cookie the host must set on ``begin_login`` and read on ``complete_login``.
    COOKIE_NAME = "gh_login"
    #: Cookie the host reads for the deep-link return target (set by the web filter).
    RETURN_COOKIE_NAME = "gh_return"
    #: Suggested ``Max-Age`` (seconds) for the ``gh_login`` cookie.
    COOKIE_MAX_AGE_SECONDS = 600

    def __init__(
        self,
        sphinx_base_url: str,
        client_id: str,
        redirect_uri: str,
        scope: str,
        signing_key: bytes,
        client: SphinxClient,
    ) -> None:
        base = sphinx_base_url[:-1] if sphinx_base_url.endswith("/") else sphinx_base_url
        require_https_or_loopback(base, "Sphinx base URL")
        self._authorize_url = base + "/oauth/authorize"
        self._client_id = client_id
        self._redirect_uri = redirect_uri
        self._scope = scope
        self._signing_key = bytes(signing_key)
        self._client = client

    def begin_login(self) -> tuple[str, str]:
        """Start a login: return ``(authorize_url, cookie_value)``. Redirect the
        browser to ``authorize_url`` and set the ``gh_login`` cookie to
        ``cookie_value`` (HttpOnly/Secure/SameSite=Lax)."""
        verifier = _b64url(os.urandom(64))
        challenge = self.challenge_for(verifier)
        query = urllib.parse.urlencode(
            {
                "response_type": "code",
                "client_id": self._client_id,
                "redirect_uri": self._redirect_uri,
                "scope": self._scope,
                "code_challenge": challenge,
                "code_challenge_method": "S256",
            }
        )
        # No state parameter — PKCE is the CSRF binding.
        return self._authorize_url + "?" + query, self.sign(verifier)

    def complete_login(
        self, gh_login_cookie: str | None, code: str | None
    ) -> TokenResponse:
        """Require this browser's verifier cookie, then redeem the code with it.

        Raises :class:`LoginCsrfError` if the cookie is absent/forged or the code
        is missing. The host should clear the ``gh_login`` cookie and rotate the
        session id (anti-fixation) after a successful call."""
        verifier = self.verify_cookie_value(gh_login_cookie)
        if verifier is None:
            raise LoginCsrfError("no login in progress for this browser")
        if code is None or not code.strip():
            raise LoginCsrfError("callback is missing the authorization code")
        return self._client.exchange_code(code, self._redirect_uri, verifier)

    def consume_return_to(
        self, gh_return_cookie: str | None, default_home: str
    ) -> str:
        """Return an open-redirect-safe same-origin path from the ``gh_return``
        cookie, or *default_home*. The host should also clear the cookie."""
        safe = self.sanitize_return_to(gh_return_cookie)
        return safe if safe is not None else default_home

    # ---- PKCE + signed cookie ("verifier.hmac") ---------------------------

    def challenge_for(self, verifier: str) -> str:
        """RFC 7636 S256: ``BASE64URL(SHA256(ASCII(verifier)))``."""
        return _b64url(hashlib.sha256(verifier.encode("ascii")).digest())

    def sign(self, verifier: str) -> str:
        return verifier + "." + _b64url(self._hmac(verifier))

    def verify_cookie_value(self, raw: str | None) -> str | None:
        """Return the verifier if the signed cookie is authentic, else ``None``."""
        if raw is None:
            return None
        dot = raw.rfind(".")
        if dot <= 0:
            return None
        verifier = raw[:dot]
        mac = raw[dot + 1 :]
        expected = _b64url(self._hmac(verifier))
        return verifier if hmac.compare_digest(mac, expected) else None

    def _hmac(self, verifier: str) -> bytes:
        return hmac.new(
            self._signing_key, verifier.encode("ascii"), hashlib.sha256
        ).digest()

    @staticmethod
    def sanitize_return_to(raw: str | None) -> str | None:
        """Return *raw* if it is a safe same-origin relative path, else ``None``."""
        if not raw:
            return None
        # Must be an absolute-path reference: exactly one leading '/'.
        if raw[0] != "/":
            return None
        if len(raw) >= 2 and raw[1] in ("/", "\\"):  # //host or /\host
            return None
        for ch in raw:
            if ord(ch) <= 0x20 or ord(ch) == 0x7F or ch == "\\":
                return None
        if "://" in raw:
            return None
        return raw

File: sdk-rust/src/login_flow.rs
//! Login-CSRF-safe hosted-login flow for Sphinx (mirrors Java `LoginFlow`).
//!
//! Binds the authorization code to the browser that started the flow using
//! **PKCE** (not `state`). Framework-agnostic: returns cookie *values* for the
//! host to set/read with its own web framework. Set the `gh_login` cookie
//! `HttpOnly`, `Secure`, `SameSite=Lax`, `Max-Age = COOKIE_MAX_AGE_SECONDS`; on
//! `complete_login` clear it and rotate the session id on elevation.

use ring::rand::{SecureRandom, SystemRandom};
use ring::{digest, hmac};

use crate::error::{LoginCsrfError, LoginError};
use crate::secure_urls;
use crate::sphinx_client::{form_component, SphinxClient, TokenResponse};

/// Cookie the host sets on `begin_login` and reads on `complete_login`.
pub const COOKIE_NAME: &str = "gh_login";
/// Cookie the host reads for the deep-link return target (set by the web filter).
pub const RETURN_COOKIE_NAME: &str = "gh_return";
/// Suggested `Max-Age` (seconds) for the `gh_login` cookie.
pub const COOKIE_MAX_AGE_SECONDS: u64 = 600;

pub struct LoginFlow {
    authorize_url: String,
    client_id: String,
    redirect_uri: String,
    scope: String,
    signing_key: Vec<u8>,
    client: SphinxClient,
}

impl LoginFlow {
    pub fn new(
        sphinx_base_url: impl Into<String>,
        client_id: impl Into<String>,
        redirect_uri: impl Into<String>,
        scope: impl Into<String>,
        signing_key: &[u8],
        client: SphinxClient,
    ) -> Self {
        let mut base = sphinx_base_url.into();
        if base.ends_with('/') {
            base.pop();
        }
        secure_urls::require_https_or_loopback(&base, "Sphinx base URL");
        Self {
            authorize_url: format!("{base}/oauth/authorize"),
            client_id: client_id.into(),
            redirect_uri: redirect_uri.into(),
            scope: scope.into(),
            signing_key: signing_key.to_vec(),
            client,
        }
    }

    /// Start a login: returns `(authorize_url, cookie_value)`. Redirect the
    /// browser to `authorize_url` and set the `gh_login` cookie to
    /// `cookie_value` (HttpOnly/Secure/SameSite=Lax).
    pub fn begin_login(&self) -> (String, String) {
        let mut buf = [0u8; 64];
        SystemRandom::new()
            .fill(&mut buf)
            .expect("CSPRNG failure generating PKCE verifier");
        let verifier = b64url(&buf);
        let challenge = self.challenge_for(&verifier);
        let query = [
            ("response_type", "code"),
            ("client_id", self.client_id.as_str()),
            ("redirect_uri", self.redirect_uri.as_str()),
            ("scope", self.scope.as_str()),
            ("code_challenge", challenge.as_str()),
            ("code_challenge_method", "S256"),
        ]
        .iter()
        .map(|&(k, v)| format!("{k}={}", form_component(v)))
        .collect::<Vec<_>>()
        .join("&");
        // No state parameter — PKCE is the CSRF binding.
        (format!("{}?{}", self.authorize_url, query), self.sign(&verifier))
    }

    /// Require this browser's verifier cookie, then redeem the code with it.
    /// The host should clear the `gh_login` cookie and rotate the session id
    /// (anti-fixation) after a successful call.
    pub fn complete_login(
        &self,
        gh_login_cookie: Option<&str>,
        code: Option<&str>,
    ) -> Result<TokenResponse, LoginError> {
        let verifier = self.verify_cookie_value(gh_login_cookie).ok_or_else(|| {
            LoginError::Csrf(LoginCsrfError {
                message: "no login in progress for this browser".into(),
            })
        })?;
        let code = match code {
            Some(c) if !c.trim().is_empty() => c,
            _ => {
                return Err(LoginError::Csrf(LoginCsrfError {
                    message: "callback is missing the authorization code".into(),
                }))
            }
        };
        self.client
            .exchange_code(code, &self.redirect_uri, Some(&verifier))
            .map_err(LoginError::Exchange)
    }

    /// Return an open-redirect-safe same-origin path from the `gh_return`
    /// cookie, or `default_home`. The host should also clear the cookie.
    pub fn consume_return_to(&self, gh_return_cookie: Option<&str>, default_home: &str) -> String {
        Self::sanitize_return_to(gh_return_cookie).unwrap_or_else(|| default_home.to_string())
    }

    // ---- PKCE + signed cookie ("verifier.hmac") ---------------------------

    /// RFC 7636 S256: `BASE64URL(SHA256(ASCII(verifier)))`.
    pub fn challenge_for(&self, verifier: &str) -> String {
        b64url(digest::digest(&digest::SHA256, verifier.as_bytes()).as_ref())
    }

    pub fn sign(&self, verifier: &str) -> String {
        format!("{verifier}.{}", b64url(self.hmac(verifier).as_ref()))
    }

    /// Return the verifier if the signed cookie is authentic, else `None`.
    pub fn verify_cookie_value(&self, raw: Option<&str>) -> Option<String> {
        let raw = raw?;
        let dot = raw.rfind('.')?;
        if dot == 0 {
            return None;
        }
        let verifier = &raw[..dot];
        let mac = &raw[dot + 1..];
        let expected = b64url(self.hmac(verifier).as_ref());
        if constant_time_eq(mac.as_bytes(), expected.as_bytes()) {
            Some(verifier.to_string())
        } else {
            None
        }
    }

    fn hmac(&self, verifier: &str) -> hmac::Tag {
        let key = hmac::Key::new(hmac::HMAC_SHA256, &self.signing_key);
        hmac::sign(&key, verifier.as_bytes())
    }

    /// Return `raw` if it is a safe same-origin relative path, else `None`.
    pub fn sanitize_return_to(raw: Option<&str>) -> Option<String> {
        let raw = raw?;
        let bytes = raw.as_bytes();
        // Must be an absolute-path reference: exactly one leading '/'.
        if bytes.first() != Some(&b'/') {
            return None;
        }
        if bytes.len() >= 2 && (bytes[1] == b'/' || bytes[1] == b'\\') {
            return None; // //host or /\host
        }
        for &b in bytes {
            if b <= 0x20 || b == 0x7f || b == b'\\' {
                return None;
            }
        }
        if raw.contains("://") {
            return None;
        }
        Some(raw.to_string())
    }
}

/// Constant-time equality for the (fixed-length) MAC comparison — no early-out
/// on content, so a forged cookie can't be discovered byte-by-byte by timing.
fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

/// URL-safe base64 without padding (RFC 4648 §5).
fn b64url(input: &[u8]) -> String {
    const A: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
    let mut out = String::with_capacity(input.len().div_ceil(3) * 4);
    for chunk in input.chunks(3) {
        let b0 = chunk[0] as u32;
        let b1 = *chunk.get(1).unwrap_or(&0) as u32;
        let b2 = *chunk.get(2).unwrap_or(&0) as u32;
        let n = (b0 << 16) | (b1 << 8) | b2;
        out.push(A[((n >> 18) & 63) as usize] as char);
        out.push(A[((n >> 12) & 63) as usize] as char);
        if chunk.len() > 1 {
            out.push(A[((n >> 6) & 63) as usize] as char);
        }
        if chunk.len() > 2 {
            out.push(A[(n & 63) as usize] as char);
        }
    }
    out
}
